python/cudf_polars/cudf_polars/experimental/statistics.py

```diff
@@ -0,0 +1,182 @@
+    IR,
+    DataFrameScan,
+    Distinct,
+    GroupBy,
+    HConcat,
+    Join,
+    Scan,
+)
+from cudf_polars.dsl.traversal import post_traversal
+from cudf_polars.experimental.base import (
+    ColumnStats,
+    StatsCollector,
+)
+from cudf_polars.experimental.dispatch import extract_base_stats
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    from cudf_polars.utils.config import ConfigOptions
+
+
+def collect_base_stats(root: IR, config_options: ConfigOptions) -> StatsCollector:
+    """
+    Collect base datasource statistics.
+
+    Parameters
+    ----------
+    root
+        Root IR node for collecting base datasource statistics.
+    config_options
+        GPUEngine configuration options.
+
+    Returns
+    -------
+    A new StatsCollector object with populated datasource statistics.
+    """
+    stats: StatsCollector = StatsCollector()
+    for node in post_traversal([root]):
+        stats.column_stats[node] = extract_base_stats(node, stats, config_options)
+    return stats
+
+
+def _update_unique_stats_columns(
+    child_column_stats: dict[str, ColumnStats],
+    key_names: Sequence[str],
+    config_options: ConfigOptions,
+) -> None:
+    """Update set of unique-stats columns in datasource."""
+    assert config_options.executor.name == "streaming", (
+        "'in-memory' executor not supported in 'add_source_stats'"
+    )
+    unique_fraction = config_options.executor.unique_fraction
+    for name in key_names:
+        if (
+            name not in unique_fraction
+            and (column_stats := child_column_stats.get(name)) is not None
+            and (source_stats := column_stats.source_info) is not None
+        ):
+            source_stats.add_unique_stats_column(column_stats.source_name or name)
+
+
+@extract_base_stats.register(IR)
+def _default_extract_base_stats(
+    ir: IR, stats: StatsCollector, config_options: ConfigOptions
+) -> dict[str, ColumnStats]:
+    # Default `extract_base_stats` implementation.
+    if len(ir.children) == 1:
+        (child,) = ir.children
+        child_column_stats = stats.column_stats.get(child, {})
+        return {
+            name: child_column_stats.get(name, ColumnStats(name=name)).copy()
+            for name in ir.schema
+        }
+    else:
+        # Multi-child nodes loose all information by default.
+        return {name: ColumnStats(name=name) for name in ir.schema}
+
+
+@extract_base_stats.register(Distinct)
+def _(
+    ir: Distinct, stats: StatsCollector, config_options: ConfigOptions
+) -> dict[str, ColumnStats]:
+    # Use default extract_base_stats after updating
+    # the known unique-stats columns.
+    (child,) = ir.children
+    child_column_stats = stats.column_stats.get(child, {})
+    key_names = ir.subset or ir.schema
+    _update_unique_stats_columns(child_column_stats, list(key_names), config_options)
+    return _default_extract_base_stats(ir, stats, config_options)
+
+
+@extract_base_stats.register(Join)
```

Comment:
Could you walk me through how this works / comment on whether this is correct? (ideally as as a comment / docstring):

The goal is to get the statistics for this `Join` node, which are determined by its input `left` and `right` children. In general, we just copy the stats over from the child notes, but there are a few special cases to consider:

1. If a column name appears in both sides of the join, we take it from the "primary" column (`right` for "Right" joins, `left` for all other joins)
2. If a column name doesn't appear in either child, it was (probably?) created during the join thanks to a `suffix`. So we copy over the stats for the column with that suffix removed.

Reply:
Yes, this is the intention right now. I added a comment.